### PR TITLE
GitLab 13.0.14 -> 13.6.0

### PR DESCRIPTION
**Pull request by jslight90**
###### Motivation for this change

Upgrade GitLab to latest version

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Besides the version changes for GitLab and its dependencies, the other changes included are:
- GitLab package:
  - update patch files for changes to GitLab code:
    - [gitlab/remove-hardcoded-locations.patch](https://github.com/NixOS/nixpkgs/pull/99116/files#diff-55b1972d258e2de5928511e45c3ddba7)
    - [gitlab/gitaly/fix-executable-check.patch](https://github.com/NixOS/nixpkgs/pull/99116/files#diff-dfffdf744fb57ef418c3f37d6da17f61)
    - [gitlab/gitlab-shell/remove-hardcoded-locations.patch](https://github.com/NixOS/nixpkgs/pull/99116/files#diff-1d9e427ec27348a3f8cd52043ec8819c)
  - remove duplicate Go dependencies from vendor directories before build:
    - [gitlab/gitaly/default.nix](https://github.com/NixOS/nixpkgs/pull/99116/files#diff-a7780142640d1079f5033c815094551cL47-R47)
    - [gitlab/gitlab-shell/default.nix](https://github.com/NixOS/nixpkgs/pull/99116/files#diff-4cb81df288767c7a74685cbd0303c4e9R21)
  - add dependency for openssl gem
    - [gitlab/default.nix](https://github.com/NixOS/nixpkgs/pull/99116/files#diff-0c101fc5429e2575d64c90df35442981)
  - add missing new line at end of data.json file:
    - [gitlab/update.py](https://github.com/NixOS/nixpkgs/pull/99116/files#diff-e8f8af88c7b2537de05e9e8bbcd945ccR120)
- NixOS module:
  - [update gitaly.toml configuration file](https://github.com/NixOS/nixpkgs/pull/99116/files#diff-a8187202c46aeec728e5b8818fdce3c8R46-R52)
  - [set path for new required Kubernetes Agent Server secret](https://github.com/NixOS/nixpkgs/pull/99116/files#diff-a8187202c46aeec728e5b8818fdce3c8R121)
    - randomly generated if non-existent
    - overwritable for actual use (either manually or with something like NixOps deployment keys)
  - [enable btree_gist extension for PostgreSQL](https://github.com/NixOS/nixpkgs/pull/99116/files#diff-a8187202c46aeec728e5b8818fdce3c8R658)

### CODE REVIEW

File: pkgs/applications/version-management/gitlab/default.nix [hunk 1]
```diff
@@ -28,6 +28,10 @@ let
           patches = [ ./fix-grpc-ar.patch ];
           dontBuild = false;
         };
+        # the openssl needs the openssl include files
```
Comment (lheckemann) [inline pkgs/applications/version-management/gitlab/default.nix:31]: Should this go in the commit with the manual changes?

Comment (lheckemann) [inline pkgs/applications/version-management/gitlab/default.nix:31]: Should this go in the commit with the manual changes?

**Comment by talyz** [comment]
@GrahamcOfBorg test gitlab

**Review by lheckemann — COMMENTED** [review]
Looks good for the most part, will test.

File: pkgs/applications/version-management/gitlab/default.nix (hunk 1 above)
Comment (jslight90) [reply pkgs/applications/version-management/gitlab/default.nix:31]: I don't see why; it's a required dependency for building the packages, as is everything else in the first commit.  The second commit adds the necessary changes to the NixOS module.

**Review by lheckemann — APPROVED** [review]
Works :+1:

**Comment by petabyteboy** [comment]
Thanks! Bumped to 13.4.3, now testing.

**Comment by petabyteboy** [comment]
After updating from 13.0.14 I had to clean `/var/gitlab/state/config` and possibly `/run/gitlab/shell-config.yml` and run `systemd-tmpfiles --create` again, otherwise the secrets would be out of sync and every push would return a 401 unauthorized. It seems to work fine now, but it would be good to make sure the upgrade path works without manual intervention.

**Comment by lheckemann** [comment]
@petabyteboy how did you "clean" `/var/gitlab/state/config`?

**Comment by lheckemann** [comment]
In my case, restarting gitaly after starting gitlab seems to have done the trick — but it occurs again after the machine is rebooted. I suspect this is because the secret is [generated at gitlab's startup](https://github.com/NixOS/nixpkgs/blob/a4bbb8a5afda47496be6d37a0b81cb25f054eb09/nixos/modules/services/misc/gitlab.nix#L856) and changes every time. I suppose gitaly should only be started after it's generated in this case — I'll test if adding `after = ["gitlab.service"];` to gitaly is sufficient for this.

**Comment by bgamari** [comment]
I can confirm that simply restarting `gitaly` after `gitlab` has started is sufficient to avoid the issue.

**Comment by petabyteboy** [comment]
I propose to merge this into unstable, create a tracking issue for the remaining known issue(s?) and possibly show a warning during evaluation to make users aware of it. When the issue is solved we can backport it as well.

**Comment by lheckemann** [comment]
I'm not keen on breaking gitlab on master, even with a warning. This will, AFAIK, _always_ require manual intervention (including after reboots and such), and not be obvious until a user complains about not being able to push.

**Comment by lheckemann** [comment]
I'm looking at fleshing out the gitlab nixos test a little, so we have a reliable way of testing an eventual fix.

**Comment by dpausp** [comment]
Gitaly should start after Gitlab and Gitaly should be restarted when Gitlab restarts, so I changed the deps like this:

~~~
systemd.services.gitaly = {
  after = [ "network.target" "gitlab.service" ];
  requires = [ "gitlab.service" ];
~~~

and 

~~~
systemd.services.gitlab = {
  after = [ "gitlab-workhorse.service" "network.target" "gitlab-postgresql.service" "redis.service" ];
  requires = [ "gitlab-sidekiq.service" ];
~~~

With this, pushing still worked after multiple restarts of the gitlab service.

**Comment by mohe2015** [comment]
I don't know the procedures for such big packages but what about updating to 13.6.0?

**Comment by petabyteboy** [comment]
I was running 13.5.x locally, 13.6.0 will require some manual work because they added a ruby dependency that tries to download stuff during the build process >.>

**Comment by petabyteboy** [comment]
I pushed an update to 13.6.0 and the fix proposed by @dpausp

**Review by ghost — APPROVED** [review]

**Comment by petabyteboy** [comment]
@GrahamcOfBorg eval

**Review by mohe2015 — APPROVED** [review]
**Only used a test install** but the upgrade worked for me.

**Comment by flokli** [comment]
Can someone take care of a backport PR for 20.09 at least?

**Comment by petabyteboy** [comment]
Thanks for merging. I will do a PR for the backport.

**Pull request merged**